test/e2e/dr/restore-from-snapshot.go

```diff
@@ -0,0 +1,342 @@
+		for _, master := range masters {
+			if master == firstMaster {
+				continue
+			}
+			runViaBastionSSH(firstMaster, proxy,
+				fmt.Sprintf("scp -o StrictHostKeyChecking=no /tmp/snapshot.db core@%s:/tmp/snapshot.db", master))
+		}
+
+		etcdConnectionString := constructEtcdConnectionString(masters, proxy)
+		e2e.Logf("etcd connstring: '%s'", etcdConnectionString)
+		for _, master := range masters {
+			runViaBastionSSH(master, proxy,
+				fmt.Sprintf("sudo -i /bin/bash -x /usr/local/bin/etcd-snapshot-restore.sh /tmp/snapshot.db %s", etcdConnectionString))
+		}
+
+		waitForAPIServer(oc)
+		waitForMastersToUpdate(oc, mcps)
+
+		rollBackInMC := getRollbackContentsInMachineConfig(oc, mc, rollBackMachineConfig)
+		o.Expect(rollBackInMC).To(o.BeEquivalentTo("data:,A"))
+
+		for _, master := range masters {
+			rollBackFile := fetchRollbackFileContents(master, proxy)
+			o.Expect(rollBackFile).To(o.BeEquivalentTo("A"))
+		}
+	})
+})
+
+func setupSSHBastion(oc *exutil.CLI) string {
+	e2e.Logf("Setting up ssh bastion host")
+	const (
+		ns = "ssh-bastion"
+	)
+
+	var (
+		bastionHost       = ""
+		sshBastionBaseDir = exutil.FixturePath("testdata", "disaster-recovery", "ssh-bastion")
+		files             = []string{
+			"service.yaml",
+			"serviceaccount.yaml",
+			"role.yaml",
+			"rolebinding.yaml",
+			"clusterrole.yaml",
+			"clusterrolebinding.yaml",
+			"deployment.yaml",
+		}
+		keyTypes = []string{"rsa", "ecdsa", "ed25519"}
+		tmpFiles = make([]string, len(keyTypes))
+	)
+
+	_, err := oc.AdminProjectClient().Project().Projects().Get(ns, metav1.GetOptions{})
+	if err != nil {
+		err = oc.Run("new-project").Args(ns).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+	}
+
+	e2e.Logf("Creating ssh keys")
+	_, err = oc.AdminKubeClient().CoreV1().Secrets(ns).Get("ssh-host-keys", metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		tmpDir, err := ioutil.TempDir("/tmp", "ssh-keys")
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer os.RemoveAll(tmpDir)
+
+		for index, keyType := range keyTypes {
+			keyPath := filepath.Join(tmpDir, keyType)
+			e2e.Logf("Generating %s key in %s", keyType, keyPath)
+			out, err := exec.Command(
+				"ssh-keygen",
+				"-q",          // silence
+				"-t", keyType, // type
+				"-f", keyPath, // output file
+				"-C", "", // no comment
+				"-N", "", // no passphrase
+			).Output()
+			if err != nil {
+				e2e.Logf("ssh-keygen output:\n%s", out)
+			}
+			o.Expect(err).NotTo(o.HaveOccurred())
+			tmpFiles[index] = keyPath
+		}
+
+		secretKeyArgs := fmt.Sprintf(
+			"ssh_host_rsa_key=%s,ssh_host_ecdsa_key=%s,ssh_host_ed25519_key=%s,sshd_config=%s",
+			tmpFiles[0], tmpFiles[1], tmpFiles[2], filepath.Join(sshBastionBaseDir, "sshd_config"),
+		)
+		_, err = oc.Run("create").Args("-n", ns, "secret", "generic", "ssh-host-keys", "--from-file", secretKeyArgs).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+	} else {
+		o.Expect(err).NotTo(o.HaveOccurred())
```

Comment:
Regarding the scope of err just making sure you are intended to report on err from

`_, err = oc.AdminKubeClient().CoreV1().Secrets(ns).Get("ssh-host-keys", metav1.GetOptions{})` on line 137

Reply:
Yeah, that's intended. "not found" error would be skipped and the secret would be created, thus two expects in different scopes.

Perhaps in-scope `err` should be renamed for clarity, wdyt?